### ISSUE TITLE
Release 4.5.0 - AWS Module testing

### ISSUE THREAD

**Issue by fdalmaup** [wazuh/wazuh#14410]
# AWS module testing

## Modules tests information
|                                      |                                                |
|--------------------------------------|------------------------------------------------|
| **Main module tests issue**          | #14376                                         |
| **Version**                          | 4.5.0                                          |
| **Release candidate #**              |                                          |
| **Tag**                              |  |
| **Previous modules tests issue**     |                                          |

## Test report procedure

All test results must have one the following statuses: 
|                 |                                                                         |
|-----------------|-------------------------------------------------------------------------|
| :green_circle: | The test passed successfully.                                            |
| :yellow_circle: | The test passed but some enhancements need to be applied to the module. |
| :red_circle:    | The test failed.                                                        |
| :white_circle:  | The test does not apply.                                                |

Any failing test must be properly addressed with a new issue, detailing the error and the possible cause. It must be included in the [Release 4.5.0 project](https://github.com/orgs/wazuh/projects/18/) for triage.

Any expected fail or skipped test must have an issue justifying the reason. All auditors must validate the justification for an expected fail or skipped test.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

Every test will be performed on a Manager unless otherwise specified.

## Test results
Python 3.9 testing:
| Bucket/Service type | Result | Issue ref. |
|---------------------|--------|------------|
| Config              | :green_circle:  | -          |
| Umbrella            | :green_circle: | -          |
| Server Access       |  :green_circle: | -          |
| CloudTrail          | :green_circle: | -          |
| VPC                 | :green_circle:   | -          |
| ALB                 | :green_circle: | -          |
| CLB                 | :green_circle:| -          |
| NLB                 | :green_circle: | -          |
| KMS               | :green_circle:| -          |
| Macie               | :green_circle:| -          |
| Trusted Advisor     | :green_circle: | -          |
| GuardDuty           | :green_circle: | -          |
| WAF                 | :green_circle:  | -          |
| Inspector           | :green_circle:  | -          |
| CloudWatch Logs | :red_circle:   | #14465          |
| Agent test          | :green_circle:  | -          |

Python 3.9 testing:
| Test                            | Status using Python 3.9 | Issue ref. |
|------------------------------|---------------------------------|--------------|
| Test using ossec.conf |  :green_circle: | - |

## Auditors validation
The definition of done for this one is the validation of the conclusions and the test results from all auditors.

All checks from below must be accepted in order to close this issue.

- [x] @davidjiglesias 
- [x] @CarlosRS9 
- [x] @fdalmaup

**Comment by fdalmaup** [comment]
## Tests performed using Python 3.9

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Bucket/Service type | Result | Issue ref. |
|---------------------|--------|------------|
| Config              |  | -          |
| Umbrella            |  | -          |
| Server Access       | | -          |
| CloudTrail          |  | -          |
| VPC                 |  | -          |
| ALB                 |  | -          |
| CLB                 | | -          |
| NLB                 |  | -          |
| Macie               | | -          |
| Trusted Advisor     |  | -          |
| GuardDuty           |  | -          |
| WAF                 |  | -          |
| Inspector           | | -          |
| Agent test          |  | -          |



## Tests performed

### Config

### Umbrella

### Server Access

### CloudTrail

###  VPCFlow

###  Test ALB

### Test CLB

### Test NLB

### Test Macie

### Test Trusted Advisor

### Test GuardDuty

### Test WAF

### Test Inspector 

## Test the module in an agent

Deploy an agent and install the required dependencies described in the official Wazuh documentation, then run the module adding the following configuration to the ossec.conf:

<details>
    <summary>Configuration</summary>

    <wodle name="aws-s3">
        <disabled>no</disabled>
        <remove_from_bucket>no</remove_from_bucket>
        <interval>10m</interval>
        <run_on_start>yes</run_on_start>
        <skip_on_error>no</skip_on_error>

        <bucket type="cloudtrail">
            <aws_profile>dev</aws_profile>
            <name>wazuh-aws-wodle-cloudtrail</name>
            <only_logs_after>2021-NOV-21</only_logs_after>
        </bucket>

        <bucket type="custom">
            <aws_profile>dev</aws_profile>
            <name>wazuh-aws-wodle-macie</name>
            <only_logs_after>2021-NOV-10</only_logs_after>
        </bucket>

        <service type="cloudwatchlogs">
            <aws_profile>dev</aws_profile>
            <aws_log_groups>4_3_test</aws_log_groups>
            <only_logs_after>2021-AUG-20</only_logs_after>
            <regions>us-east-1</regions>
        </service>
    </wodle>
</details>

**Comment by fdalmaup** [comment]
Hi team!

Some of the expected results don't match with the actual ones, like the VPCFlow tests and the **Execute the module having an only_logs_after later than the last key** test of the Config section.

I'll inspect the underlying code to determine what could be the problem.

**Comment by fdalmaup** [comment]
KMS
CloudWatch Logs
ECR Image scanning


Test log 0 - 2022-7-29
Test log 1
Test log 2
Test log 3

# CloudWatch Logs

## Test details
Before starting these tests, a series of Cloudwatch Logs test events in a period of 5 days in the past until today must be uploaded to the `4_5_test`'s `test_stream` log stream.

### Test only_logs_after

Remove the `/var/ossec/wodles/aws/aws_services.db` file and run the module using the following command to obtain every log from the Cloudwatch Logs log group for the given date.

#### First execution without specifying an only_logs_after value

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

    /var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -p dev -r us-east-1
    DEBUG: +++ Debug mode on - Level: 2
    DEBUG: +++ Getting alerts from "us-east-1" region.
    DEBUG: only logs: None
    DEBUG: +++ Table does not exist; create
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: Getting data from DB for log stream "test_stream" in log group "4_5_test"
    DEBUG: Token: "None", start_time: "None", end_time: "None"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "None", start_time "1659063600000" and end_time "None"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 0 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 1 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 2 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 3 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 4 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 5 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 6 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 7 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 8 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 9 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 10 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 11 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 12 - 2022-7-29"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "f/36999179178242526779471046389549471113305287701024210944/s", start_time "1659063600000" and end_time "None"
    DEBUG: Saving data for log group "4_5_test" and log stream "test_stream".
    DEBUG: The saved values are "{'token': 'f/36999179178275977895263112801254887538255401511053918207/s', 'start_time': 1659063600000, 'end_time': 1659100574840}"
    DEBUG: Purging the BD
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: committing changes and closing the DB


</details>

##### Conclusions
The module only collected the logs uploaded the date of execution so it is working as expected.

#### First execution specifying an only_logs_after value set in the past
Remove the database file and run the command specifying a date set 5 days earlier in the past. All the logs between the date set by the `only_logs_after` parameter and the present day must be collected.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

    /var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s 2022-JUL-24 -p dev -r us-east-1
    DEBUG: +++ Debug mode on - Level: 2
    DEBUG: +++ Getting alerts from "us-east-1" region.
    DEBUG: only logs: 1658620800000
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: Getting data from DB for log stream "test_stream" in log group "4_5_test"
    DEBUG: Token: "None", start_time: "None", end_time: "None"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "None", start_time "1658620800000" and end_time "None"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log event 2022-07-26"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 0 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 1 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 2 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 3 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 4 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 5 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 6 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 7 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 8 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 9 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 10 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 11 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 12 - 2022-7-29"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "f/36999179178242526779471046389549471113305287701024210944/s", start_time "1658620800000" and end_time "None"
    DEBUG: Saving data for log group "4_5_test" and log stream "test_stream".
    DEBUG: The saved values are "{'token': 'f/36999179178275977895263112801254887538255401511053918207/s', 'start_time': 1658620800000, 'end_time': 1659100574840}"
    DEBUG: Some data already exists on DB for that key. Updating their values...
    DEBUG: Purging the BD
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
DEBUG: committing changes and closing the DB

</details>

Keep the database file and run the same command again to ensure no duplicate logs are processed. Use the same `only_logs_after` parameter.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

    /var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s 2022-JUL-24 -p dev -r us-east-1
    DEBUG: +++ Debug mode on - Level: 2
    DEBUG: +++ Getting alerts from "us-east-1" region.
    DEBUG: only logs: 1658620800000
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: Getting data from DB for log stream "test_stream" in log group "4_5_test"
    DEBUG: Token: "None", start_time: "None", end_time: "None"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "None", start_time "1658620800000" and end_time "None"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log event 2022-07-26"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 0 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 1 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 2 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 3 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 4 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 5 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 6 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 7 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 8 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 9 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 10 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 11 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 12 - 2022-7-29"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "f/36999179178242526779471046389549471113305287701024210944/s", start_time "1658620800000" and end_time "None"
    DEBUG: Saving data for log group "4_5_test" and log stream "test_stream".
    DEBUG: The saved values are "{'token': 'f/36999179178275977895263112801254887538255401511053918207/s', 'start_time': 1658620800000, 'end_time': 1659100574840}"
    DEBUG: Some data already exists on DB for that key. Updating their values...
    DEBUG: Purging the BD
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: committing changes and closing the DB

</details>




#### Execute the module having a last key set and no only_logs_after value
Without removing the database file, execute `sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET end_time = XXXXXXXXXXXXX, next_token=null;'`, being `XXXXXXXXXXXXX` the timestamp in milliseconds corresponding to the fourth log.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -p dev -r us-east-1
```
The module should collect all the logs starting from the `Test log 5`.

<details>
    <summary>Command output</summary>

    sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET end_time = 1659100492369, next_token=null;'
    /var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -p dev -r us-east-1
    DEBUG: +++ Debug mode on - Level: 2
    DEBUG: +++ Getting alerts from "us-east-1" region.
    DEBUG: only logs: None
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: Getting data from DB for log stream "test_stream" in log group "4_5_test"
    DEBUG: Token: "None", start_time: "None", end_time: "None"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "None", start_time "1659063600000" and end_time "None"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 0 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 1 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 2 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 3 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 4 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 5 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 6 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 7 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 8 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 9 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 10 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 11 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 12 - 2022-7-29"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "f/36999179178242526779471046389549471113305287701024210944/s", start_time "1659063600000" and end_time "None"
    DEBUG: Saving data for log group "4_5_test" and log stream "test_stream".
    DEBUG: The saved values are "{'token': 'f/36999179178275977895263112801254887538255401511053918207/s', 'start_time': 1659063600000, 'end_time': 1659100574840}"
    DEBUG: Some data already exists on DB for that key. Updating their values...
    DEBUG: Purging the BD
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: committing changes and closing the DB


</details>



#### Execute the module having an only_logs_after later than the last key
Keeping the database file, execute `sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET end_time = XXXXXXXXXXXXX next_token=null;'`, being `XXXXXXXXXXXXX` the timestamp in milliseconds corresponding to the fourth log.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```
Being `YYYY-MMM-DD` the date corresponding to the day following the execution date. No log should be collected.

<details>
    <summary>Command output</summary>

    sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET end_time = 1659100492369, next_token=null;'
	/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s 2022-JUL-30 -p dev -r us-east-1
    DEBUG: +++ Debug mode on - Level: 2
    DEBUG: +++ Getting alerts from "us-east-1" region.
    DEBUG: only logs: 1659139200000
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: Getting data from DB for log stream "test_stream" in log group "4_5_test"
    DEBUG: Token: "None", start_time: "None", end_time: "None"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "None", start_time "1659139200000" and end_time "None"
    DEBUG: Saving data for log group "4_5_test" and log stream "test_stream".
    DEBUG: The saved values are "{'token': 'f/36999764967047712198546403386786436463350019838189731839/s', 'start_time': 1659139200000, 'end_time': 1659139200000}"
    DEBUG: Some data already exists on DB for that key. Updating their values...
    DEBUG: Purging the BD
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: committing changes and closing the DB

</details>




#### Execute the module having an only_logs_after earlier than the last key
Keeping the database file, execute `sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET start_time = XXXXXXXXXXXXX, end_time = YYYYYYYYYYYYY, next_token=null;'`, being `XXXXXXXXXXXXX` the timestamp in milliseconds corresponding to the fourth log, and `YYYYYYYYYYYYY` the timestamp in milliseconds corresponding to the sixth log. Then run the module specifying the day before the date of execution as `only_logs_after`. The module should collect all logs in the log stream except for those with a timestamp between `YYYYYYYYYYYYY` and `XXXXXXXXXXXXX`.


Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

    sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET start_time = 1659100492369, end_time = 1659100512690, next_token=null;'
    /var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s 2022-JUL-28 -p dev -r us-east-1
    DEBUG: +++ Debug mode on - Level: 2
    DEBUG: +++ Getting alerts from "us-east-1" region.
    DEBUG: only logs: 1658966400000
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: Getting data from DB for log stream "test_stream" in log group "4_5_test"
    DEBUG: Token: "None", start_time: "None", end_time: "None"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "None", start_time "1658966400000" and end_time "None"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 0 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 1 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 2 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 3 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 4 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 5 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 6 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 7 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 8 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 9 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 10 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 11 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 12 - 2022-7-29"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "f/36999179178242526779471046389549471113305287701024210944/s", start_time "1658966400000" and end_time "None"
    DEBUG: Saving data for log group "4_5_test" and log stream "test_stream".
    DEBUG: The saved values are "{'token': 'f/36999179178275977895263112801254887538255401511053918207/s', 'start_time': 1658966400000, 'end_time': 1659100574840}"
    DEBUG: Some data already exists on DB for that key. Updating their values...
    DEBUG: Purging the BD
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: committing changes and closing the DB



</details>

Then run the same command again to ensure that the module is not processing duplicate events.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

	/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_5_test -d 2 -s 2022-JUL-28 -p dev -r us-east-1
    DEBUG: +++ Debug mode on - Level: 2
    DEBUG: +++ Getting alerts from "us-east-1" region.
    DEBUG: only logs: 1658966400000
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: Getting data from DB for log stream "test_stream" in log group "4_5_test"
    DEBUG: Token: "None", start_time: "None", end_time: "None"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "None", start_time "1658966400000" and end_time "None"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 0 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 1 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 2 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 3 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 4 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 5 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 6 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 7 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 8 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 9 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 10 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 11 - 2022-7-29"
    DEBUG: +++ Sending events to Analysisd...
    DEBUG: The message is "Test log 12 - 2022-7-29"
    DEBUG: Getting CloudWatch logs from log stream "test_stream" in log group "4_5_test" using token "f/36999179178242526779471046389549471113305287701024210944/s", start_time "1658966400000" and end_time "None"
    DEBUG: Saving data for log group "4_5_test" and log stream "test_stream".
    DEBUG: The saved values are "{'token': 'f/36999179178275977895263112801254887538255401511053918207/s', 'start_time': 1658966400000, 'end_time': 1659100574840}"
    DEBUG: Some data already exists on DB for that key. Updating their values...
    DEBUG: Purging the BD
    DEBUG: Getting log streams for "4_5_test" log group
    DEBUG: Found "test_stream" log stream in 4_5_test
    DEBUG: committing changes and closing the DB
 

</details>
